pilz_control/test/integrationtest_pilz_joint_trajectory_controller.py

```diff
@@ -121,155 +120,197 @@ def observeStop(self, max_deceleration = MAX_DECELERATION, timeout = STOP_DURATI
+        return self._monitored_cartesian_speed_srv.turn_off_speed_monitoring()
+
+    def _turn_on_speed_monitoring(self):
+        return self._monitored_cartesian_speed_srv.turn_on_speed_monitoring()
+
+    def test_hold_at_startup(self):
+        """Tests that 'hold' mode is active at startup of controller."""
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=1,
+                                      expected_error_code=FollowJointTrajectoryResult.INVALID_GOAL),
+                        "Motion did not fail although controller should be in 'hold' mode at startup")
+
+        self.assertTrue(self._hold_srv.request_default_mode(), 'Switch to default mode failed')
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=0.5),
+                        "Motion failed although controller should be in 'unhold' mode")
+
+    def test_hold_during_motion(self):
+        """Activate hold mode while robot is moving and evaluate executed stop."""
+        is_executing_srv = IsExecutingServiceWrapper()
+
+        self._start_motion(position = 0.2, duration = 5)
 
         # Wait for movement to commence
-        MovementObserver().observeUntilPositionGreater(0.11, 3)
-        self.assertTrue(is_executing_srv.call(), 'is_executing service returned false.')
-        # Hold robot during movement
-        rospy.loginfo("  - Robot is moving -> Switch into HOLDING mode")
-        self.assertTrue(hold_srv.requestHoldingMode(), 'Switch to Mode HOLDING failed.')
+        MovementObserver().observe_until_position_greater(0.11, 3)
+        self.assertTrue(is_executing_srv.call(), "Controller is not executing")
+        rospy.loginfo("Robot is moving -> Switch into HOLDING mode")
+        self.assertTrue(self._hold_srv.request_holding_mode(), 'Switch to Mode HOLDING failed.')
 
         # Make sure robot stops (not abruptly)
-        MovementObserver().observeStop()
-        rospy.loginfo("  - Stop of robot observed!")
+        MovementObserver().observe_stop()
+        rospy.loginfo("Stop of robot observed!")
+
+        self.assertTrue(self._wait_for_motion_result(), "Motion failed")
+
+        self.assertEqual(GoalStatus.PREEMPTED, self._trajectory_dispatcher.get_last_state(), 'Goal was not preempted.')
+
+    def test_speed_monitoring(self):
+        """Tests if controller detects Cartesian speed limit violation and switches to hold mode."""
+        far_away_position = 0.5
+
+        self.assertTrue(self._turn_on_speed_monitoring(), "Could not turn on speed monitoring")
+
+        self.assertTrue(self._hold_srv.request_default_mode(), "Switch to default mode failed")
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=0.5),
+                        "Motion to default position failed")
+
+        self.assertTrue(self._move_to(position=far_away_position, duration=0.1),
+                        "Cartesian speed monitor did not detect speed limit violation")
```

Comment:
This checks if the result is `SUCCESSFUL`, so you should use `assertFalse`.

Reply:
No. This is the problem: Even so the motion is cancelled, the goal still returns `SUCCESSFUL` as `error_code`, therefore, I check for `true` here. This problem is due to a strange implementation in the goal. We talked about this. I can, however, add a comment.

Reply:
Ok. I did not see that.
In this case I would separate the `_move_to` call and the assertion such that the latter can be commented out and add a small comment.

Reply:
Done.